### ISSUE TITLE
Async never completes in this scenario (parent/child)

### ISSUE THREAD

**Issue by venkatperi** [Kotlin/kotlinx.coroutines#308]
Given two `Deferred`s `a1` and `a2` with `a2` waiting on `a1`, specifying `a1` as the parent of `a2` causes `a2` to never complete:

```
val a1 = CompletableDeferred<Int>()
val a2 = async(a1) {
   a1.await() + 1
 }

a1.complete(1)
runBlocking {
   a2.await()
}
```

Works fine when without the parent relationship i.e. `val a2 = async {...}` 

What am I missing here? Possible "deadlock" with parent relationship specified + awaiting on a1?

**Comment by jcornaz** [comment]
Parents always wait on their children. So parents always complete after their children. So yes if a child is waiting for its parent, it will never complete.

**Comment by venkatperi** [comment]
Looks like a classic case of deadlock. Closing.
